### ISSUE TITLE
[IMPROVEMENT] The replica scheduler may consider more factors

### ISSUE THREAD

**Issue by Vicente-Cheng** [longhorn/longhorn#5149]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

The replica scheduler may consider more aspects, like replica counts and the scheduled size.
We may schedule the replica to the same disk if other volumes are not used because of the thin-provision design. We calculate the disk usage only if allocated.
We could consider more when replica scheduling to improve data balance and performance.
Also, it would benefit the sharding implementation. (If sharding is implemented, the balance of whole disks on the nodes would be important)

## Describe the solution you'd like

TBD with HEP

## Additional context

The example for current behavior.
Nodes A, B, and C all have two disks. 
  have -> disk A (200G), disk B (150 G)

We would schedule the replica to disk A until the whole volume size is bigger than the disk size * overcommit ratio instead of scheduling more balance.

**Comment by innobead** [comment]
cc @longhorn/dev

**Comment by derekbit** [comment]
Ref: https://github.com/longhorn/longhorn/issues/4230

**Comment by innobead** [comment]
> Ref: [[FEATURE] Different replica scheduling strategies](https://github.com/longhorn/longhorn/issues/4230)

Let's consolidate both issues, and use this one instead.

**Comment by innobead** [comment]
ref: https://github.com/longhorn/longhorn/issues/4613

**Comment by innobead** [comment]
Can consider node/disk anti-affinity for replica scheduling.

**Comment by Vicente-Cheng** [comment]
> Can consider node/disk anti-affinity for replica scheduling or zone affinity.

Did you mean that we could group the node/disk for replica scheduling (like whitelist) or blacklist for node/disk anti-affinity?

**Comment by joshimoo** [comment]
Consider doing the below as part of the replica scheduler improvement.
https://github.com/longhorn/longhorn/issues/4826

**Comment by iosifnicolae2** [comment]
it would be great to have two soft anti-affinity options:
- `Replica Node Level Soft Anti-Affinity` - to schedule replicas on different nodes
- `Replica Disk Level Soft Anti-Affinity` - to schedule replicas on different disks

For example, on a single node Kubernetes cluster where HA is not required but data loss is a problem we would disable the above two anti-affinity options and we would get a software RAID based on Longhorn.
  - the advantage of using Longhorn for implementing the software RAID is that we can add a few more nodes and in a few clicks we get an HA cluster
 
Slack discussion: https://rancher-users.slack.com/archives/CC2UQM49Y/p1684827748167229

**Comment by iosifnicolae2** [comment]
### Configure volume replicas using Persistent Volume Claim
It would be extremly helpful to be able to configure the number of replicas using Persistent Volume Claim labels.

Currently, we have multiple Storage Classes with default replica count, but when we want to change the number of replicas programatically, we do the change manually from UI and this is not convinient to automate.

**Comment by ejweber** [comment]
> It would be extremly helpful to be able to configure the number of replicas using Persistent Volume Claim labels.
> 
> Currently, we have multiple Storage Classes with default replica count, but when we want to change the number of replicas programatically, we do the change manually from UI and this is not convinient to automate.

You might consider modifying the spec of the Longhorn volume for an easily programmatic way of modifying the replica count. For example:

```
eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get volume -o jsonpath='{.spec.numberOfReplicas}{"\n"}' test
2
eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get replicas
NAME              STATE     NODE                                DISK                                   INSTANCEMANAGER                                     IMAGE                                    AGE
test-r-a0c5ef24   running   eweber-v124-worker-1ae51dbb-4pngn   9699a051-aff8-457d-a6f4-4c067615e7ed   instance-manager-bd54850e35239b37069374780ca0f9a3   longhornio/longhorn-engine:master-head   15s
test-r-a29095b9   running   eweber-v124-worker-1ae51dbb-pbxr9   4a758df2-1b96-4c11-a664-4f211f542923   instance-manager-4639bda14281d41f3af00d64bc364bb9   longhornio/longhorn-engine:master-head   15s
eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system patch volume test --type merge -p '{"spec": {"numberOfReplicas": 3}}'
volume.longhorn.io/test patched
eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get volume -o jsonpath='{.spec.numberOfReplicas}{"\n"}' test
3
eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get replicas
NAME              STATE     NODE                                DISK                                   INSTANCEMANAGER                                     IMAGE                                    AGE
test-r-033140d7   running   eweber-v124-worker-1ae51dbb-ppvzp   711d4bcb-85ab-4cfc-ad06-fcf12de42916   instance-manager-faf47d6949e508720ae71ed3ee10e466   longhornio/longhorn-engine:master-head   13s
test-r-a0c5ef24   running   eweber-v124-worker-1ae51dbb-4pngn   9699a051-aff8-457d-a6f4-4c067615e7ed   instance-manager-bd54850e35239b37069374780ca0f9a3   longhornio/longhorn-engine:master-head   36s
test-r-a29095b9   running   eweber-v124-worker-1ae51dbb-pbxr9   4a758df2-1b96-4c11-a664-4f211f542923   instance-manager-4639bda14281d41f3af00d64bc364bb9   longhornio/longhorn-engine:master-head   36s
```

**Comment by iosifnicolae2** [comment]
> > It would be extremly helpful to be able to configure the number of replicas using Persistent Volume Claim labels.
> > 
> > Currently, we have multiple Storage Classes with default replica count, but when we want to change the number of replicas programatically, we do the change manually from UI and this is not convinient to automate.
> 
> You might consider modifying the spec of the Longhorn volume for an easily programmatic way of modifying the replica count. For example:
> 
> ```
> eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get volume -o jsonpath='{.spec.numberOfReplicas}{"\n"}' test
> 2
> eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get replicas
> NAME              STATE     NODE                                DISK                                   INSTANCEMANAGER                                     IMAGE                                    AGE
> test-r-a0c5ef24   running   eweber-v124-worker-1ae51dbb-4pngn   9699a051-aff8-457d-a6f4-4c067615e7ed   instance-manager-bd54850e35239b37069374780ca0f9a3   longhornio/longhorn-engine:master-head   15s
> test-r-a29095b9   running   eweber-v124-worker-1ae51dbb-pbxr9   4a758df2-1b96-4c11-a664-4f211f542923   instance-manager-4639bda14281d41f3af00d64bc364bb9   longhornio/longhorn-engine:master-head   15s
> eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system patch volume test --type merge -p '{"spec": {"numberOfReplicas": 3}}'
> volume.longhorn.io/test patched
> eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get volume -o jsonpath='{.spec.numberOfReplicas}{"\n"}' test
> 3
> eweber@laptop:~/longhorn-manager> kubectl -n longhorn-system get replicas
> NAME              STATE     NODE                                DISK                                   INSTANCEMANAGER                                     IMAGE                                    AGE
> test-r-033140d7   running   eweber-v124-worker-1ae51dbb-ppvzp   711d4bcb-85ab-4cfc-ad06-fcf12de42916   instance-manager-faf47d6949e508720ae71ed3ee10e466   longhornio/longhorn-engine:master-head   13s
> test-r-a0c5ef24   running   eweber-v124-worker-1ae51dbb-4pngn   9699a051-aff8-457d-a6f4-4c067615e7ed   instance-manager-bd54850e35239b37069374780ca0f9a3   longhornio/longhorn-engine:master-head   36s
> test-r-a29095b9   running   eweber-v124-worker-1ae51dbb-pbxr9   4a758df2-1b96-4c11-a664-4f211f542923   instance-manager-4639bda14281d41f3af00d64bc364bb9   longhornio/longhorn-engine:master-head   36s
> ```

hmm, got it, but we're deploying our application using helm charts, so we define just the PVC and not the PV

But for now, this is an acceptable solution as we can implement some hooks to update PV replica count.

Thank you!

**Comment by ejweber** [comment]
> hmm, got it, but we're deploying our application using helm charts, so we define just the PVC and not the PV

Got it. So, just to be clear, you're saying you want to:

- Deploy the application with a PVC that has some number of replicas based on the StorageClass (e.g. 3).
- Later, modify the deployment so that a different number of replicas are used (e.g. 2).
- But your automation only knows about the PVC (not the underlying Longhorn volume), so you want to manipulate it directly?

**Comment by iosifnicolae2** [comment]
> > hmm, got it, but we're deploying our application using helm charts, so we define just the PVC and not the PV
> 
> Got it. So, just to be clear, you're saying you want to:
> 
> * Deploy the application with a PVC that has some number of replicas based on the StorageClass (e.g. 3).
> * Later, modify the deployment so that a different number of replicas are used (e.g. 2).
> * But your automation only knows about the PVC (not the underlying Longhorn volume), so you want to manipulate it directly?

Yes, we want to control the number of replicas from PVC (the number of replicas from StorageClass will be used if no replica count label is not defined on PVC).

**Comment by innobead** [comment]
@ChanYiLin Please help with this. You can discuss this with @Vicente-Cheng because he has some ideas about this.

**Comment by innobead** [comment]
> Yes, we want to control the number of replicas from PVC (the number of replicas from StorageClass will be used if no replica count label is not defined on PVC).

@c3y1huang this is similar to recurring job applied to PVC instead of PV.
